api/src/main/java/org/apache/iceberg/actions/RewriteTablePath.java

```diff
@@ -0,0 +1,102 @@
+ *   <li><b>Incremental copy:</b> Rewrites a subset of metadata files to the staging directory,
+ *       consisting of metadata files added since a specified start version and/or until end
+ *       version. The start/end version is identified by the name of a metadata.json file, and all
+ *       metadata files added before/after these file are marked for rewrite.
+ * </ul>
+ *
+ * This action can be used as the starting point to fully or incrementally copy an Iceberg table
+ * located under the source prefix to the target prefix.
+ *
+ * <p>The action returns the following:
+ *
+ * <ol>
+ *   <li>The name of the latest metadata.json file copied.
+ *   <li>A listing of rewritten table metadata files, relative to the staging directory, that can be
+ *       copied to the target prefix.
+ *   <li>A listing of table data files, relative to the source prefix, that can be copied to the
+ *       target prefix.
+ * </ol>
+ */
+public interface RewriteTablePath extends Action<RewriteTablePath, RewriteTablePath.Result> {
+
+  /**
+   * Configure a source prefix that will be replaced by the specified target prefix in all paths
+   *
+   * @param sourcePrefix the source prefix to be replaced
+   * @param targetPrefix the target prefix
+   * @return this for method chaining
+   */
+  RewriteTablePath rewriteLocationPrefix(String sourcePrefix, String targetPrefix);
+
+  /**
+   * First metadata version to rewrite, identified by name of a metadata.json file in the table's
+   * metadata log. It is optional, if provided then this action will only rewrite metadata files
+   * added after this version.
+   *
+   * @param startVersion name of a metadata.json file. For example,
+   *     "00001-8893aa9e-f92e-4443-80e7-cfa42238a654.metadata.json".
+   * @return this for method chaining
+   */
+  RewriteTablePath startVersion(String startVersion);
+
+  /**
+   * Last metadata version to rewrite, identified by name of a metadata.json file in the table's
+   * metadata log. It is optional, if provided then this action will only rewrite metadata files
+   * added before this file, including the file itself.
+   *
+   * @param endVersion name of a metadata.json file. For example,
+   *     "00001-8893aa9e-f92e-4443-80e7-cfa42238a654.metadata.json".
+   * @return this for method chaining
+   */
+  RewriteTablePath endVersion(String endVersion);
+
+  /**
+   * Custom staging location. It is optional. By default, staging location is a subdirectory under
+   * table's metadata directory.
+   *
+   * @param stagingLocation the staging location
+   * @return this for method chaining
+   */
+  RewriteTablePath stagingLocation(String stagingLocation);
+
+  /** The action result that contains a summary of the execution. */
+  interface Result {
+    /** staging location of rewritten files */
+    String stagingLocation();
+
+    /**
+     * path to a list of comma-separated pairs of source and target path for data and metadata
```

Comment:
How about:

Path to a list of comma-separated pairs of source path to target path for all files in the table added between startVersion and endVersion.  Note that this list includes data files from the original table and metadata files rewritten to staging.

Reply:
Also tagging @KolonelKustard who is planning to consume this functionality if any thoughts on whether this is nice to consume for the copy part.

Reply:
+1

Having a single list means that as the table format evolves and the implementations of this copy action evolve, hopefully there's fewer changes for downstream users of this action.

Having the target path saves downstream users having to duplicate the prefix re-writing rules already applied by this actions implementations. Which hopefully makes using this a bit less error-prone.